tools-local/ensure-nupkg-dependencies-on-source/Program.cs

```diff
@@ -0,0 +1,193 @@
+
+using System.CommandLine;
+using ensure_nupkg_dependencies_on_source;
+using NuGet.Common;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+internal class Program
+{
+    private static async Task<int> Main(string[] args)
+    {
+        var rootCommand = new RootCommand("Check that nupkg dependencies are available on specified source(s).");
+
+        var nupkgsArgument = new Argument<List<FileInfo>>("nupkgs");
+        nupkgsArgument.Arity = ArgumentArity.OneOrMore;
+        nupkgsArgument.Description = "Nupkgs to check";
+        rootCommand.Add(nupkgsArgument);
+
+        var sourcesOption = new Option<List<string>>("--source");
+        sourcesOption.AddAlias("-s");
+        sourcesOption.IsRequired = true;
+        sourcesOption.Arity = ArgumentArity.OneOrMore;
+        rootCommand.Add(sourcesOption);
+
+        rootCommand.SetHandler(ExecuteAsync, nupkgsArgument, sourcesOption);
+        var exitCode = await rootCommand.InvokeAsync(args);
+        return exitCode;
+    }
+
+    private static async Task<int> ExecuteAsync(List<FileInfo> files, List<string> sourcesList)
+    {
+        if (!CheckAllFilesExist(files))
+        {
+            return -1;
+        }
+
+        IReadOnlyList<SourceRepository> sources = GetSources(sourcesList);
+        List<string> messages = new();
+        IReadOnlyDictionary<string, PackageInfo> nupkgs = GetNupkgInfo(files, messages);
+
+        await CheckDependenciesExistAsync(nupkgs, sources, messages);
+
+        if (messages.Count == 0)
+        {
+            Console.WriteLine("No missing dependencies found");
+            return 0;
+        }
+        else
+        {
+            foreach (var message in messages)
+            {
+                Console.WriteLine(message);
+            }
+            return -1;
+        }
+    }
+
+    private static async Task CheckDependenciesExistAsync(
+        IReadOnlyDictionary<string, PackageInfo> nupkgs,
+        IReadOnlyList<SourceRepository> sources,
+        List<string> messages)
+    {
+        Dictionary<PackageIdentity, bool> checkedPackages = new();
+
+        // The packages provided are intended to be pushed, so pretend they're already there
+        foreach (var packageInfo in nupkgs.Values)
+        {
+            checkedPackages.Add(packageInfo.PackageIdentity, true);
+        }
+
+        using (var cacheContext = new SourceCacheContext())
+        {
+            foreach (var (nupkgPath, packageInfo) in nupkgs)
+            {
+                foreach (var dependency in packageInfo.Dependencies)
+                {
+                    bool exists = false;
+                    if (!checkedPackages.TryGetValue(dependency, out exists))
+                    {
+                        foreach (var source in sources)
+                        {
+                            var resource = await source.GetResourceAsync<FindPackageByIdResource>();
```

Comment:
I'd probably generate this at line 76. You don't need to get the resource constantly.